### ISSUE TITLE
Suggest replacing .map(f).flatten() with .and_then(f) on Options

### ISSUE THREAD

**Issue by scottmcm** [rust-lang/rust-clippy#6870]
### What it does

*What does this lint do?*
Checks for usage of `_.map(_).flatten()`,

This is like https://rust-lang.github.io/rust-clippy/master/index.html#map_flatten but for `Option`s instead of `Iterator`s

### Categories (optional)

- Kind: clippy::style/complexity

Easier to follow what's happening, shorter

### Drawbacks

None.

### Example

Inspired by @steffahn's suggestion in https://github.com/Nadrieril/dhall-rust/pull/213#issuecomment-792793983

**Comment by giraffate** [comment]
Do you mean this case? <https://github.com/rust-lang/rust-clippy/blob/4c8d248190819f09753d4f6a9f89e3804e232ae7/tests/ui/map_flatten.stderr#L33-L39>

**Comment by scottmcm** [comment]
Ah, so I did!  The description in the docs made me think it's only for `Iterator::map`.

**Comment by steffahn** [comment]
Which means the description should be fixed, right?

**Comment by scottmcm** [comment]
Thanks, @giraffate!

**Comment by jyn514** [comment]
Hmm, this doesn't seem to be working: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=d9855ae1539cc222b88fa4baa7b46b92
```rust
    let _: Option<&str> = x.as_ref().map(|pb| pb.to_str()).flatten();
```
doesn't get a warning.

**Comment by steffahn** [comment]
@jyn514 It’d a pedantic lint; not enabled by default.

https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=1e75f19dfbb4f5b35b2f2ef4a9d8b200

**Comment by jyn514** [comment]
Ah ok, that makes sense. What do you think about making it a style lint instead?

**Comment by steffahn** [comment]
Seems reasonable to me to have it be a style lint. I’m not really sure I understand the distinction anyways; I see a lot of similar types of lints on both sides of the distinction, e.g. [`option_map_or_none`](https://rust-lang.github.io/rust-clippy/master/index.html#option_map_or_none) is a style lint that also suggests `and_then` for a different type of pattern, and there are lots of iterator-related style lints for cases where a combination of iterator methods can be simplified, e.g. [`iter_skip_next`](https://rust-lang.github.io/rust-clippy/master/index.html#iter_skip_next).

More pedantic lints other than `map_flatten` that seem similar in spirit to other style lints include [`filter_map_next`](https://rust-lang.github.io/rust-clippy/master/index.html#filter_map_next) and [`manual_ok_or`](https://rust-lang.github.io/rust-clippy/master/index.html#manual_ok_or).

**Comment by orhun** [comment]
I think this should be a style lint too. I didn't know about this lint until I figure out I can use `and_then` instead of `map` & `flatten` so it would be good to be warned about this by `clippy`, as default.
